### PR TITLE
Add iterator support to HNSW index

### PR DESCRIPTION
**Pull request by zhengbuqian**
This PR adds iterator support to HNSW index. Issue: #61.

* Add iterator interface in IndexNode
* Add iterator support to HNSW index
* Add `seed_ef` config
* Add an integration test that simulates kNN using iterator

/kind improvement

### CODE REVIEW

File: thirdparty/hnswlib/hnswlib/hnswalg.h [hunk 1]
```diff
@@ -335,62 +391,13 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
         // the connectivity of the graph; but we do not want to keep all of them as they won't be candidates. Thus we
         // include only a subset of filtered nodes(controlled by kAlpha) in the search path.
         float accumulative_alpha = 0.0f;
+        auto add_search_candidate = [&](Neighbor n) { return retset.insert(n, disqualified); };
```
Comment (foxspy) [inline thirdparty/hnswlib/hnswlib/hnswalg.h:394]: confused here. I see iterator use the `disqualified` to hold the candidates, but the retset seems never be used in iterator, am I right? If so, the cost of insert operation for retset can be reduced.

**Comment by foxspy** [comment]
A cool work~

File: thirdparty/hnswlib/hnswlib/hnswalg.h (hunk 1 above)
Comment (zhengbuqian) [reply thirdparty/hnswlib/hnswlib/hnswalg.h:394]: for iterator, this retset is the initial search result using seed_ef and all points in `retset` are later added to `workspace->retset`. 

points in `retset`/`workspace->retset` are those ready to be returned by `iterator->Next()` and won't be searched in later iteration phase; points in `disqualified`/`workspace->to_visit` still need to be searched before they can possibly be returned by `iterator->Next()`.

**Review by foxspy — APPROVED** [review]

File: thirdparty/hnswlib/hnswlib/hnswalg.h [hunk 2]
```diff
@@ -1269,6 +1269,67 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
...
+            }
+            for (int i = 0; i < retset.size(); i++) {
+                workspace->retset.push(retset[i]);
+            }
+            workspace->initial_search_done = true;
+        }
+        if (!workspace->to_visit.empty()) {
+            auto top = workspace->to_visit.top();
+            workspace->to_visit.pop();
+            auto add_search_candidate = [&](Neighbor n) {
```
[the remark this answers is not in the document]
Comment (zhengbuqian) [reply thirdparty/hnswlib/hnswlib/hnswalg.h:1310]: Good catch! For connectivity we need to add those filtered points to workspace->to_visit but we don't want to add them to workspace->retset.

**Pull request merged**